File: LeetCode/Java/Jingshen/99_Recover_Binary_Search_Tree.java
/**
Two elements of a binary search tree (BST) are swapped by mistake.

Recover the tree without changing its structure.

Note:
A solution using O(n) space is pretty straight forward. Could you devise a constant space solution?

Jingshen
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void recoverTree(TreeNode root) {
        List<TreeNode> swap=new ArrayList<TreeNode>();
        swap.add(new TreeNode(Integer.MIN_VALUE));
        helper(root,swap);
        if(swap.size()==3)
        {
            int tmp=swap.get(0).val;
            swap.get(0).val=swap.get(1).val;
            swap.get(1).val=tmp;
        }
        else
        {
            int tmp=swap.get(0).val;
            swap.get(0).val=swap.get(3).val;
            swap.get(3).val=tmp;
        }
    }
    private void helper(TreeNode root, List<TreeNode> swap) {
        if(root==null)
            return;
        helper(root.left,swap);
        if(swap.get(swap.size()-1).val<root.val)
            swap.set(swap.size()-1,root);
        else
        {
            swap.add(root);
            swap.add(root);
        }
        helper(root.right,swap);
    }
}

File: LeetCode/C++/99_Recover_Binary_Search_Tree.cpp
//99. Recover Binary Search Tree
/*
Two elements of a binary search tree (BST) are swapped by mistake.
Recover the tree without changing its structure.

Tag: Tree, Depth-first Search

Author: Xinyu Liu
*/
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private: 

    void inorder_trans(TreeNode* root, TreeNode*& pre, TreeNode*& swap1, TreeNode*& swap2){
        if(!root) return;
        inorder_trans(root->left, pre, swap1, swap2);
        
        if(swap1 == NULL && root->val < pre->val){
            swap1 = pre;
        }
        if(swap1 != NULL && root->val < pre->val){
            swap2 = root;
        }
        pre = root;
        inorder_trans(root->right, pre, swap1, swap2);
    }
public:
    void recoverTree(TreeNode* root) {
        if(!root) return;
        TreeNode* pre = new TreeNode(INT_MIN);
        TreeNode* swap1 = NULL;
        TreeNode* swap2 = NULL;
        inorder_trans(root, pre, swap1, swap2);
        int tmp = swap1->val;
        swap1->val = swap2->val;
        swap2->val = tmp;
    }
};

void main(){
    TreeNode n1(2),n2(1),n3(3);
    n1.left = &n2;
    n2.right = &n3;
    Solution sol;
    sol.recoverTree(&n1);
}
